internal/auth/options.go

```diff
@@ -129,20 +131,82 @@ type SignatureData struct {
 	key  string
 }
 
-// NewOptions returns new options
-func NewOptions() *Options {
-	return &Options{
-		Port:            4180,
-		CookieName:      "_sso_auth",
-		CookieSecure:    true,
-		CookieHTTPOnly:  true,
-		CookieExpire:    time.Duration(168) * time.Hour,
-		CookieRefresh:   time.Duration(0),
-		SetXAuthRequest: false,
-		PassUserHeaders: true,
-		PassHostHeader:  true,
-		ApprovalPrompt:  "force",
-		RequestLogging:  true,
+// NewOptions returns new options with the below overrides
+func NewOptions() (*Options, error) {
+	overrides := map[string]interface{}{
+		"port":             4180,
+		"cookie_name":      "_sso_auth",
+		"cookie_secure":    true,
+		"cookie_http_only": true,
+		"cookie_expire":    "168h",
+		"cookie_refresh":   "1h",
+		"set_xauthrequest": false,
+		"passUser_headers": true,
+		"passHost_header":  true,
+		"approval_prompt":  "force",
+		"request_logging":  true,
+	}
+	options, err := loadVars(overrides)
+	if err != nil {
+		return nil, err
+	}
+	return options, nil
+}
+
+func loadVars(overrides map[string]interface{}) (*Options, error) {
+	var opts Options
+
+	bindAllOptVars(reflect.TypeOf(&opts).Elem(), "mapstructure")
+	setDefaults()
+
+	for key, value := range overrides {
+		viper.Set(key, value)
+	}
+
+	err := viper.Unmarshal(&opts)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode env vars into options struct")
+	}
+	return &opts, nil
+}
+
+// bindAllOptVars takes in a struct with tags and uses the tag values to bind env vars
+func bindAllOptVars(t reflect.Type, tag string) error {
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		tagValue := field.Tag.Get(tag)
+		err := viper.BindEnv(tagValue)
+		if err != nil {
+			return fmt.Errorf("Unable to bind env var: %q", tagValue)
+		}
+	}
+	return nil
+}
+
+// setDefaults sets config defaults for the default viper instance
+func setDefaults() {
```

Comment:
my gut is to say we should take a `v *viper.Viper` here and use that to set defaults

Reply:
👍 Creating a new viper instance in `NewOptions` and passing it around to relevant functions.